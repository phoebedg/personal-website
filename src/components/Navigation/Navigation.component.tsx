import React, { useState } from "react";
import { Link, useLocation } from "react-router-dom";
import threeLines from "../../static/threeLines.svg";
import "./Navigation.css";

const navMenuItems = ["home", "personal", "professional", "projects", "press"];

export const Navigation = () => {
  const [clicked, setClicked] = useState<boolean>(false);
  const toggleNav = (e: React.MouseEvent) => {
    e.preventDefault();
    setClicked(!clicked);
  };
  const closeNav = (e: React.MouseEvent) => {
    setClicked(false);
  };

  const location = useLocation();
  const getStyles = (link: string): string =>
    location.pathname === `/${link}`
      ? "navigation__link active"
      : "navigation__link";

  return (
    <div className="navigation">
      <img
        className="navigation__icon"
        src={threeLines}
        alt="nav icon"
        onClick={toggleNav}
      />
      {clicked && (
        <div className="navigation__modal" onClick={closeNav}>
          <div className="navigation__menu">
            {navMenuItems.map(item => {
              return (
                <Link
                  className={getStyles(item)}
                  onClick={closeNav}
                  to={`/${item}`}
                  key={item}
                >
                  {item}
                </Link>
              );
            })}
          </div>
        </div>
      )}
    </div>
  );
};
