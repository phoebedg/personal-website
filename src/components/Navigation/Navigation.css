.navigation {
  display: flex;
  flex-direction: row;
  padding: 20px;
}

.navigation__icon {
  width: 20px;
  cursor: pointer;
}

.navigation__modal {
  position: fixed;
  top: 20px;
  left: 40px;
  width: 100%;
  height: 100%;
  background: initial;
}

.navigation__menu {
  display: flex;
  justify-content: space-evenly;
  min-width: 270px;
  width: 450px;
}

.navigation__link {
  text-decoration: none;
  color: #d1d7c9;
  font-weight: 800;
}

.navigation__link:hover,
.navigation__link:focus,
.navigation__link:active,
.navigation__link:visited,
.active {
  text-decoration: underline;
}
