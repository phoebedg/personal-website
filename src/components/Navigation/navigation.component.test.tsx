import React from "react";
import { BrowserRouter as Router, useLocation } from "react-router-dom";
import { render, fireEvent } from "@testing-library/react";
import { Navigation, navMenuItems } from "./Navigation.component";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(() => ({
    pathname: "/"
  }))
}));

test("renders Navigation icon", async () => {
  const { container, getByRole } = render(
    <Router>
      <Navigation />
    </Router>
  );
  expect(getByRole("img")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("clicking nav icon opens nav modal menu", () => {
  const { getByRole, getAllByTestId } = render(
    <Router>
      <Navigation />
    </Router>
  );
  fireEvent.click(getByRole("img"));
  const navLinks = getAllByTestId("nav-link");
  expect(navLinks).toBeTruthy();
  expect(navLinks.map(l => l.innerHTML)).toEqual(navMenuItems.map(i => i));
  expect(navLinks[0].className).toEqual("navigation__link active");
});

test("renders nav modal menu", () => {
  const { container, getByRole } = render(
    <Router>
      <Navigation />
    </Router>
  );
  fireEvent.click(getByRole("img"));
  expect(container).toMatchSnapshot();
});

// TODO: fix test
test("clicking nav link closes nav modal menu", () => {
  (useLocation as jest.Mock).mockImplementation(() => ({
    pathname: "/personal"
  }));
  const { getByRole, getByText } = render(
    <Router>
      <Navigation />
    </Router>
  );
  fireEvent.click(getByRole("img"));
  expect(getByText("personal").className).toEqual("navigation__link active");
});
