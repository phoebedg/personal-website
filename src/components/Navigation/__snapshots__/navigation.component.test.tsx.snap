// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Navigation icon 1`] = `
<div>
  <div
    class="navigation"
  >
    <img
      alt="nav icon"
      class="navigation__icon"
      src="threeLines.svg"
    />
  </div>
</div>
`;

exports[`renders nav modal menu 1`] = `
<div>
  <div
    class="navigation navigation-mobile-open"
  >
    <img
      alt="nav icon"
      class="navigation__icon"
      src="threeLines.svg"
    />
    <div
      class="navigation__modal"
    >
      <div
        class="navigation__menu"
      >
        <a
          class="navigation__link active"
          data-testid="nav-link"
          href="/"
        >
          home
        </a>
        <a
          class="navigation__link"
          data-testid="nav-link"
          href="/professional/#main"
        >
          professional
        </a>
        <a
          class="navigation__link"
          data-testid="nav-link"
          href="/personal/#main"
        >
          personal
        </a>
        <a
          class="navigation__link"
          data-testid="nav-link"
          href="/projects/#main"
        >
          projects
        </a>
        <a
          class="navigation__link"
          data-testid="nav-link"
          href="/press/#main"
        >
          press
        </a>
      </div>
    </div>
  </div>
</div>
`;
